### ISSUE TITLE
io_uring new uSockets backend

### ISSUE THREAD

**Issue by uNetworkingAB** [uNetworking/uWebSockets#1603]
As of Linux 5.19 I now see slight outperformance with io_uring for an actual networked ping/ping benchmark involving client and server TCP processes. It's less than 10% on my computers, often less than 5%, so we do not see any gigantic gains since we come from a highly optimized epoll usage. But there are gains, esp. for mitigated systems.

Long term plan is to add a new backend in uSockets. It's very, very low priority but I'm interested long term

**Comment by uNetworkingAB** [comment]
Scratch that - as of Linux 6.0 there are significant gains with io_uring. I can see 18% gains in testing if using the 6.0 features, and they make a lot of sense now even for low memory cases and many restrictions have been removed.

This is going to be a new backend in uSockets, probably the default one as soon as possible. I like it and it will be nice to have a path that is entirely optimized solely for Linux.

**Comment by uNetworkingAB** [comment]
Scratch that, the gains I'm seeing is more like 21%

**Comment by uNetworkingAB** [comment]
Linux 6.0+ is crazy fast. I have a build of uWS that does HTTP messaging with URL router over io_uring, faster than raw TCP messaging over epoll.

This perf. is going to compound for pub/sub use cases and there's no doubt that io_uring will be default in v21. This is a game changer.

My shitty 10 year old semi-budget PC can do 406k messages per second, on 1 CPU core. This was 325k with epoll. And it does 336k URL routed HTTP messages per second. That's slap in the face kind of performance.

**Comment by dalisoft** [comment]
@uNetworkingAB  Is there will be release of **uWebSockets.js** as alpha for this implementation?

**Comment by uNetworkingAB** [comment]
I was still experimenting with a POC 5 days ago, and got this working in uSockets for the first time yesterday. So we are many months away from even beginning to think about what to do about Node.js integration.

But there has always been a difference between what is used in uWebSockets.js and what is used in uWebSockets - one use libuv and the other used raw epoll. So it will most likely just change to being libuv for one and io_uring for the other.

**Comment by dalisoft** [comment]
So this implementation only for C++?

**Comment by uNetworkingAB** [comment]
It will be in Bun. Node.js - I don't know. Depends on how extensible libuv is but I doubt it.

**Comment by angelsanzn** [comment]
For the sake of completeness: apparently libuv is working on its own implementation of io_uring as a "backend". First proposed in 2018 (https://github.com/libuv/libuv/issues/1947) but movement started very recently (https://github.com/libuv/libuv/pull/3952, https://github.com/libuv/libuv/pull/3979).

**Comment by uNetworkingAB** [comment]
Oh ow, I'm up almost 100k messages per second:

Messages per second: 424716.812500
Messages per second: 425812.031250
Messages per second: 423331.312500
Messages per second: 424703.812500

This used to be 325k or something

**Comment by uNetworkingAB** [comment]
Holy cock!

Req/sec: 469954.250000
Req/sec: 467404.500000
Req/sec: 468516.500000

The latest io_uring Linux "for-next" branch is speedy as all heck

**Comment by kolinfluence** [comment]
@uNetworkingAB 
just curious, do u think af_xdp will be faster?

**Comment by uNetworkingAB** [comment]
I see it as two different targets: either you say "we are making a **fast** web server in user space for **standard** Linux systems for maximum **compatibility**"

or you say "We are **bypassing** Linux network stack with various kernel space hacks for **maximum** performance".

You can always do specialized bypasses and even remove Linux altogether and just run realtime systems or ASICs for whatever you are doing, but none of that has the "vanilla compatibility" of simple user space apps. At least not as far as I know.

You can grab this compiled ELF binary and run on any Linux system.

It's also relevant to know the history: we started as a Node.js addon and has always been limited by whatever platforms Node.js supports. Only recently has the project been separated from Node.js and become its own stand-alone project and only just now are we looking at more Linux-only features like io_uring, kTLS.

So I think those are two different goals.

**Comment by uNetworkingAB** [comment]
Yeah reading up on this AF_XDP is really just a newer variant of AF_PACKET and is just raw packet processing, bypassing the built-in network stack of Linux.

That's not where I'm interested in going - I want to use Linux, not avoid it. So io_uring with Linux networking is still a huge improvement compared to epoll with Linux networking (at least small messages so far).

You can find all kinds of bypasses, it's not new, but they all require depending on some third party networking stack and then you have a whole mountain of integration problems. You're balancing compatibility and ease of use against balls to the walls performance extremism :D

I don't mind it, but it's not something mainstream ppl are going to use.

**Comment by uNetworkingAB** [comment]
Mainstream ppl want to double click Ubuntu and double click a program and have it run. And if it can run **way faster** than other double-clickable apps then that's a win, given the constraints they run with.

But if you are the very NASDAQ you probably don't scan GitHub for projects to use, then you pay people a shit ton of money, who write very, very fine tuned specialized apps only for that usage.

**Comment by e3dio** [comment]
> double click Ubuntu and double click a program and have it run

That is way too much clicking, I would prefer to only click once please, thanks. These mice don't last forever

**Comment by uNetworkingAB** [comment]
TLDR - AF_XDP is kernel bypassing, we are not doing that here. The target platform for uWS is "vanilla Linux", not bypasses.

It's also hard to tell if that's actually better than Linux networking given that it depends on what drivers your particular NIC has, such as whether it does full bypass on its own or does TLS offloading, etc.

Bypasses drastically reduce the ease of use / applicability to real world deployments in real companies.

**Comment by hiqsociety** [comment]
@uNetworkingAB 
maybe someday can do speed comparison with multicore of https://github.com/bytedance/monoio for http.
curious how much perf difference vs rust based program is

**Comment by uNetworkingAB** [comment]
I've done some thinking and investigation,

The reason why epoll is faster than io_uring for 8kb and 16kb message sending is because io_uring is utilizing much more memory (epoll uses 1 single receive buffer and 1 single send buffer, while io_uring uses per-socket receive buffers and send buffers).

With more memory usage, you won't fit in CPU-cache - which is what epoll does.

If I run 2 processes with epoll, sending TCP messages back and forth with 300k msg/sec on localhost, it only produces a fraction of that in cache misses. This means that the entire hot path of epoll benchmarking on localhost never even writes to RAM - it's entirely to/from CPU cache.

144 million TCP messages sent and received only produces 2 million cache misses. So the entire "message sending" between the two involved processes never even leaves the CPU cache. It's a total short circuit.

So benchmarking on localhost with epoll / io_uring is very misleading unless you can find a good balance between excessive io_uring calls and memory usage.

Most likely, io_uring requires a CPU with much cache - mine only has 6MB while modern CPUs have 96MB. With more cache, it will be easier to efficiently batch io_uring calls in order to get better performance.

I think that, io_uring will never perform as good as epoll on old CPUs for this reason, as there is not enough CPU cache to queue up enough calls to be faster than the equivalence of multiple calls to send syscall.

I need a better set up (hardware) to get apples-to-apples comparison between epoll and io_uring, probably a 10gbit ethernet card with 2 connectors with a wire connectin the two. That way you gurarantee both solutions must go to NIC

**Comment by billywhizz** [comment]
> I need a better set up (hardware) to get apples-to-apples comparison between epoll and io_uring, probably a 10gbit ethernet card with 2 connectors with a wire connectin the two. That way you gurarantee both solutions must go to NIC

your analysis makes sense. i have never been able to make io_uring out-perform epoll in local benches and had assumed it was just a skill issue, but maybe not.

re. the 10Gb - techempower have recently upgraded their hardware to 10Gb between the DB/load generator/Web Server, so it might be worth putting it in there with io_uring enabled and not to see if it makes a difference? I dunno if it's enabled by default in the existing node.js/uWebSockets and Bun implementations on TE.

**Comment by rafalwijata** [comment]
I cannot reproduce it with epoll backend thus I post it here only. While using uring backend, and testing HelloWorld example with
./wrk -t 8 -c 1000 -d 10s http://localhost:3000/foo
the server sometimes(quite often actually) calls any("*/") route rather than get("/*") route even though req->getMethod() returns "get" within that any() route. Any thoughts?
